File: src/main/java/akostenko/aicars/drawing/LinesBuilder.java
package akostenko.aicars.drawing;

import akostenko.aicars.math.Decart;
import akostenko.aicars.math.Polar;
import akostenko.aicars.math.Vector;

import org.newdawn.slick.Color;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

class LinesBuilder {
    private List<LocalLine> lines = new ArrayList<>();

    LinesBuilder between(Vector from, Vector to) {
        lines.add(new LocalLine(from, to));
        return this;
    }

    LineFromTo from(Vector from) {
        return new LineFromTo(from);
    }

    Collection<LocalLine> build() {
        return lines;
    }

    class LineFromTo {
        private final Vector from;
        LineFromTo(Vector from) {
            this.from = from;
        }

        LinesBuilder towards(double direction, double size) {
            LinesBuilder.this.lines.add(new LocalLine(from, from.plus(new Polar(size, direction))));
            return LinesBuilder.this;
        }
    }

    /** All this lines are correct only while image is centered at (0, 0)
    /* They are incompatible with {@link Line} while are unscaled and have no color
     */
    static class LocalLine {
        private final Vector from;
        private final Vector to;

        LocalLine(Vector from, Vector to) {
            this.from = from;
            this.to = to;
        }

        LocalLine rotate(double radians) {
            return new LocalLine(from.rotate(radians), to.rotate(radians));
        }

        LocalLine scale(Scale scale) {
            float k = scale.getPixels() / scale.getSize();
            return new LocalLine(from.multi(k), to.multi(k));
        }

        Line position(Decart position, Color color, int width) {
            return new Line(from.toDecart().plus(position),
                    to.toDecart().plus(position),
                    color,
                    width);
        }
    }
}

File: src/main/kotlin/akostenko/aicars/drawing/LinesBuilder.kt
package akostenko.aicars.drawing

import akostenko.aicars.math.Decart
import akostenko.aicars.math.Polar
import akostenko.aicars.math.Vector
import org.newdawn.slick.Color
import java.util.*

internal class LinesBuilder {
    private val lines = ArrayList<RelativeLine>()

    fun between(from: Vector, to: Vector): LinesBuilder {
        lines.add(RelativeLine(from, to))
        return this
    }

    fun from(from: Vector): LineFromTo {
        return LineFromTo(from)
    }

    fun build(): Collection<RelativeLine> {
        return lines
    }

    internal inner class LineFromTo(private val from: Vector) {

        fun towards(direction: Double, size: Double): LinesBuilder {
            this@LinesBuilder.lines.add(RelativeLine(from, from.plus(Polar(size, direction))))
            return this@LinesBuilder
        }
    }

    /** All this lines are correct only while image is centered at (0, 0)
     * They are incompatible with [StraightLine] while are unscaled and have no color
     */
    internal class RelativeLine(val from:Vector, val to:Vector) {

        fun rotate(radians:Double): RelativeLine {
            return RelativeLine(from.rotate(radians), to.rotate(radians))
        }

        fun scale(scale:Scale): RelativeLine {
            val k = scale.pixels / scale.size
            return RelativeLine(from * k, to * k)
        }

        fun place(position:Decart, color:Color, width:Float): StraightLine {
            return StraightLine(from.toDecart().plus(position),
                    to.toDecart().plus(position),
                    color,
                    width)
        }
    }

    internal class RelativeArc(val center: Vector, val radius: Double, val from: Double, val to: Double) {

        fun rotate(radians: Double): RelativeArc {
            return RelativeArc(center, radius, from + radians, to + radians)
        }

        fun scale(scale: Scale): RelativeArc {
            val k = scale.pixels / scale.size
            return RelativeArc(center * k, radius * k, from, to)
        }

        fun place(position: Decart, color:Color, width:Float): ArcLine {
            return ArcLine(position + center, radius, from, to, color, width)
        }
    }
}
